### DISCUSSION TITLE
Why DockerOperator provide logs with different header formats?

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#29302] [General]
Hello everyone,
I have my Airflow instance running on a docker-compose environment.
It spins up  a Celery worker (containerized) from where I execute some container with some python code inside of it.
The python code print some logs with the Print function and also with the logging python module.
On the interface and on the log files I can see that these logs sometimes are appended with the docker.py header and sometimes not.
Every run i make I can see a different behaviour (e.g. logs appended with docker.py header change everytime).
How is that possible?
![image](https://user-images.githubusercontent.com/89450255/216261696-dcc0b4ac-a5bf-464f-9493-d74597340245.png)

**Comment by potiuk** [comment]
I think it depends on your code in Python. Very interesting case - can you share a reproducible example/DAG that you have and container that you run ? Definitely loooks like some kind of race condition.

**Reply by MattiaGallegati** [reply]
The code is as simple as:
```
def main():
    print("TESTING SCRIP1")
    print("TESTING SCRIPT2")
    print("TESTING SCRIPT3")
    print("TESTING SCRIPT4")
    import logging
    import sys
    root = logging.getLogger()
    root.setLevel(logging.DEBUG)
    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    root.addHandler(handler)
    root.error("1")
    root.error("2")
    root.error("3")
    root.error("4")

if __name__ == "__main__":
    main()
```
Build in a container like:
```
ARG BASE_IMAGE=python:3.8-slim-buster
FROM ${BASE_IMAGE}
ENV PYTHONUNBUFFERED=1
ENV FROM_DOCKER_FILE=1

COPY ./src ./app
```
executed in a DAG like:

```
default_args = {
	'owner'                 : 'airflow',
	'description'           : 'Use of the DockerOperator',
	'depend_on_past'        : False,
	'start_date'            : datetime(2018, 1, 3),
	'email_on_failure'      : False,
	'email_on_retry'        : False
}

with DAG('docker_dag', default_args=default_args, schedule_interval="5 * * * *", catchup=False) as dag:
	t2 = DockerOperator(
	task_id='docker_command',
	image='airflow_poc:latest',
	api_version='auto',
	auto_remove=True,
	docker_url="unix://var/run/docker.sock",
	network_mode="bridge",
	command="python3 ./app/main.py"
	)

t2
```

Airflow version 2.4.2

**Comment by potiuk** [comment]
Ah it's easy. because the output of print in the docker operator is treated as single "string" - with multiple lines.  so

```
TESTING 1
TESTING 2
TESTING 3
...
```

Is part of single log entry. This is how Python logging works. You can write your own Handler if you want to handle it differently.

**Reply by MattiaGallegati** [reply]
That's maybe the case for the first logs, but as you can see in the code (and from the output) I also use a python logger that shows a quite similar behaviour. 
I can see this "problem" also in other projects, airflow version 2.5.0. 
From time to time logs are printed without headers.

**Reply by potiuk** [reply]
No. It's the same story. the logs are pulled periodically and what you print inside the docker operator, when it contains multiple log lines, they are printed as a single log entry. This is the very same story.

Again  - if you really want, you can configure your logging handlers to split multi-line message into separate log lines. This is not how Python logging works by default but if you have a need to do so, you can definitely implement your own handler.

**Reply by MattiaGallegati** [reply]
Just to see if I understood correctly, you are pointing our that: 
root.error("1")
root.error("2")
root.error("3")
root.error("4")
are multiline logs?
To my understanding (which maybe is not correct) these are separated single lines (infact you can see that each log has its own Python header), and must be filled with Airflow headers in front of it.
The fact that is missing probably is due to the "log sampling" behaviour of DockerOperator you pointed out.
Maybe, if the logs are "too near" in time, they will be sampled in a single string (so they becomes multiline).
What do you think?

**Reply by potiuk** [reply]
Yes. That'what i meant by pulling logs by Docker operator. If you want to change it. PRs are most welcome

**Comment by potiuk** [comment]
(and BTW. If you think it can be improved by default and see the value in it - this is of course most welcome if you can contribute back a change in behaviour. I can imagine that you could have a flag controlling multi-line behaviour in Airflow. That would be a nice feature (and PRs as usual are most welcome).
